src/Agent.Plugins/Artifact/PipelineArtifactServer.cs

```diff
@@ -138,121 +139,66 @@ public PipelineArtifactServer(IAppTraceSource tracer)
-                    {
-                        buildArtifact = await buildServer.GetArtifact(downloadParameters.ProjectId, downloadParameters.PipelineId, downloadParameters.ArtifactName, cancellationToken);
-                    }
-                    else if (downloadParameters.ProjectRetrievalOptions == BuildArtifactRetrievalOptions.RetrieveByProjectName)
+                    if (string.IsNullOrEmpty(downloadParameters.ProjectName))
                     {
-                        if (string.IsNullOrEmpty(downloadParameters.ProjectName))
-                        {
-                            throw new InvalidOperationException("Project name can't be empty when trying to fetch build artifacts!");
-                        }
-                        else
-                        {
-                            buildArtifact = await buildServer.GetArtifactWithProjectNameAsync(downloadParameters.ProjectName, downloadParameters.PipelineId, downloadParameters.ArtifactName, cancellationToken);
-                        }
+                        throw new InvalidOperationException("Project name can't be empty when trying to fetch build artifacts!");
                     }
                     else
                     {
-                        throw new InvalidOperationException($"Invalid {nameof(downloadParameters.ProjectRetrievalOptions)}!");
+                        artifacts = await buildServer.GetArtifactsWithProjectNameAsync(downloadParameters.ProjectName, downloadParameters.PipelineId, cancellationToken);
                     }
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Invalid {nameof(downloadParameters.ProjectRetrievalOptions)}!");
+                }
 
-                    var manifestId = DedupIdentifier.Create(buildArtifact.Resource.Data);
-                    var options = DownloadDedupManifestArtifactOptions.CreateWithManifestId(
-                        manifestId,
+                IEnumerable<BuildArtifact> pipelineArtifacts = artifacts.Where(a => string.Equals(a.Resource.Type, PipelineArtifactConstants.PipelineArtifact, StringComparison.OrdinalIgnoreCase));
+                if (!pipelineArtifacts.Any())
+                {
+                    throw new ArgumentException("Could not find any pipeline artifacts in the build.");
+                }
+                else
+                {
+                    var (dedupManifestClient, clientTelemetry) = await DedupManifestArtifactClientFactory.Instance
+                        .CreateDedupManifestClientAsync(
+                        context.IsSystemDebugTrue(),
+                        (str) => context.Output(str),
+                        connection,
+                        DedupManifestArtifactClientFactory.Instance.GetDedupStoreClientMaxParallelism(context),
+                        WellKnownDomainIds.DefaultDomainId,
```

Comment:
we should probably add support for alternate domains eventually...

Reply:
yep.